tasmota/tasmota_xdrv_driver/xdrv_92_esp32_gnss.ino

```diff
@@ -0,0 +1,876 @@
+#define D_CMND_GNSS "GNSS"
+#define RTC_UPDATE_INTERVAL (30 * 60) // Every 30 minutes
+#define INVALIDATE_AFTER 30           // 30 seconds
+#define SERIAL_INPUT_BUFFER_SIZE 2048
+#define SERIAL_STREAM_BUFFER_SIZE 4096
+
+#ifdef USE_AUX_WEBSERVER
+#define MAX_STREAMING_CLIENTS 10 // Maximum number of simultaneous streaming clients
+#define CLIENT_TIMEOUT 30000     // 30 seconds timeout for streaming clients
+#define AUX_WEBSERVER_PORT  8080
+#endif
+
+/*********************************************************************/
+/*                          STRUCTURES                               */
+/*********************************************************************/
+struct NMEACounters
+{
+  uint32_t gga;   // Global Positioning System Fix Data
+  uint32_t rmc;   // Recommended Minimum Navigation Information
+  uint32_t gll;   // Geographic Position - Latitude/Longitude
+  uint32_t gsa;   // GNSS DOP and Active Satellites
+  uint32_t gsv;   // GNSS Satellites in View
+  uint32_t vtg;   // Course Over Ground and Ground Speed
+  uint32_t other; // Any other NMEA sentences
+};
+
+struct SerialConfigMap
+{
+  const char *name;
+  uint32_t config;
+};
+
+typedef struct
+{
+  uint32_t crc32;
+  uint32_t version;
+  uint32_t baudrate;
+  uint32_t serial_config;
+} gnss_settings_t;
+
+#ifdef USE_AUX_WEBSERVER
+struct StreamClient
+{
+  AsyncClient *client;
+  uint32_t lastDataTime;
+  bool isActive;
+};
+#endif
+
+/*********************************************************************/
+/*                   CLI COMMANDS                                    */
+/*********************************************************************/
+
+void CmndGNSSBaudrate(void);
+void CmndGNSSSerialConfig(void);
+void CmndGNSSSend(void);
+
+void (*const GNSSCommand[])(void) PROGMEM = {
+  &CmndGNSSBaudrate,
+  &CmndGNSSSerialConfig,
+  &CmndGNSSSend};
+
+const char kGNSSCommands[] PROGMEM = D_CMND_GNSS "|"
+                                               "Baudrate|"     // Set UART baudrate
+                                               "SerialConfig|" // Set UART configuration
+                                               "Send";         // Send data to GNSS module
+
+/*********************************************************************/
+/*                         GLOBAL VARIABLES                          */
+/*********************************************************************/
+static gnss_settings_t GNSSSettings;
+
+static struct
+{
+  uint32_t baudrate;
+  uint32_t serial_config;
+  uint8_t valid;
+  uint32_t last_valid;
+  float latitude;
+  float longitude;
+  float altitude;
+  uint8_t satellites;
+  float hdop;
+  char time[32];
+  char date[32];
+  char last_valid_time[32];
+  NMEACounters nmea_counters;
+  uint32_t total_nmea_sentences;
+#ifdef USE_AUX_WEBSERVER
+  uint32_t stream_clients;
+#endif
+} GNSSData;
+
+static TinyGPSPlus gps;
+static GNSSParser gnssParser;
+static HardwareSerial *gnssSerial = nullptr;
+
+static uint32_t last_rtc_sync = 0;
```

Comment:
It is preferred to declare a unique 'struct' for all globals for each driver. It makes compiled code more compact and avoids name collisions